### PR TITLE
Feature/presentation signup

### PR DESCRIPTION
**Pull request by jacquesporveau**
Main features being:
 - Presentations can be created by non-admin users that are signed in.
 - Presentations are auto assigned to current user on creation
 - Event dropdown only shows upcoming events

Pivotal: https://www.pivotaltracker.com/n/projects/1976733

### CODE REVIEW

File: spec/controllers/presentations_controller_spec.rb [hunk 1]
```diff
@@ -55,29 +55,12 @@
 
     before { mock_pundit_user_as(non_admin) }
 
-    describe 'GET #new' do
```
Comment (harmonymjb) [inline spec/controllers/presentations_controller_spec.rb:58]: It may be a good idea to continue to test these controller endpoints.
Controller specs give you access to a response you can assert against.
e.g. `expect(response.status).to eq(401)` or `expect { subject }.to redirect_to(unauthorized_path)`

File: app/controllers/presentations_controller.rb [hunk 2]
```diff
@@ -9,7 +9,12 @@ def new
   end
 
   def create
-    @presentation = Presentation.new(presentation_params)
+    # because of the presentation policy we know a person
+    # will have to be signed in to view this page.
+    # current_person will never be nil
```
Comment (harmonymjb) [inline app/controllers/presentations_controller.rb]: Which page is `this page`?
Do you think the reader would expect that `current_user` _could_ be `nil`?
What is expected to happen if I POST to `/presentations` without logging in? Can this be tested for in the controller spec?

File: app/views/presentations/_form.html.erb [hunk 3]
```diff
@@ -18,11 +18,9 @@
 
   <p>
     <%= f.label :event %><br>
-    <%= collection_select(:presentation, :event_id, Event.all, :id, :name, prompt: true) %>
+    <%= collection_select(:presentation, :event_id, Event.upcoming, :id, :name, prompt: true) %>
   </p>
 
-  <%= hidden_field_tag(:current_person, current_person.id) %>
```
Comment (harmonymjb) [inline app/views/presentations/_form.html.erb]: This line seems unrelated to this commit.
If this line is not needed, the history can be rebased so that it is never added.

File: spec/controllers/presentations_controller_spec.rb (hunk 1 above)
Comment (jacquesporveau) [reply spec/controllers/presentations_controller_spec.rb:58]: Yeah the reason I took these tests out was with the presentation pundit policy changing, non-admin and admin users share the same privileges as each other when accessing the "new" and "create" controller actions. So if I were to keep these tests here I would just have to change them to be testing the exact same thing that the admin tests are testing.

Feel like if the policy is tested, which it is. We can be sure that this is working as intended.

File: app/controllers/presentations_controller.rb (hunk 2 above)
Comment (jacquesporveau) [reply app/controllers/presentations_controller.rb]: Yep that's not very clear. I meant you wouldn't be able to view `/presentations/new` without being logged in first.

`current_user` has the potential to be nil if you made the post request from like a ARC or AJAX request outside the app. In that case you would be greeted with a Pundit not authorized error. 

Which is not tested. I'll write up specs for that.

File: app/views/presentations/_form.html.erb [hunk 4]
```diff
@@ -18,14 +18,7 @@
...
     <%= f.label :event %><br>
-    <%= collection_select(:presentation, :event_id, Event.all, :id, :name, prompt: true) %>
-  </p>
-
-  <p>
-    <%= f.label :presenter %><br>
-    <%= collection_select(:presentation, :person_id, Person.all, :id, :first_name, include_blank: true) %>
-    or
-    <%= f.text_field :presenter, placeholder: 'Enter guest name' %>
+    <%= collection_select(:presentation, :event_id, Event.upcoming, :id, :name, prompt: true) %>
```
Comment (dangerdogz) [inline app/views/presentations/_form.html.erb]: This can be associated with the form object and replaced by `<%= f.collection_select(:event_id, Event.upcoming, :id, :name, prompt: true) %>`

Otherwise, the default value will not be selected from the @presentation object

Comment (fredericboivin) [inline app/views/presentations/_form.html.erb]: This can be associated with the form object and replaced by `<%= f.collection_select(:event_id, Event.upcoming, :id, :name, prompt: true) %>`

Otherwise, the default value will not be selected from the @presentation object

**Review by fredericboivin — CHANGES_REQUESTED** [review]

File: app/views/presentations/_form.html.erb (hunk 4 above)
Comment (jacquesporveau) [reply app/views/presentations/_form.html.erb]: That sounds like a good idea. Worth mentioning that as there is no edit action implemented yet I don't think there is a situation in which this form would appear that a presentation would already have an associated event already assigned.

However this form will surely be used when the edit functionality is introduced so I am going to include it. Cheers thanks!

**Review by fredericboivin — APPROVED** [review]

File: spec/controllers/presentations_controller_spec.rb (hunk 1 above)
Comment (harmonymjb) [reply spec/controllers/presentations_controller_spec.rb:58]: I think the policy test is good. It tests that the policy returning correctly.
The policy is tested in one context, i.e. `new?` returns `true` is meaningful only if you understand the context of the policy's responsibility, and that you're sure it's correctly applied to the controller.
It might seem intuitive that is `new?` returns `true` on the policy means that the controller is functioning properly, but as the complexity of the application grows, you'll want to test the the policy returns `true` and `false` correctly _and_ that your controller produces the responses (`http_status`, redirect, etc.) that you expect.

Keep in mind what the controller's responsibility is and how it is different than the policy's.
